pkg/activator/throttler.go

```diff
@@ -271,3 +295,57 @@ func (t *Throttler) endpointsDeleted(obj interface{}) {
 	revID := RevisionID{ep.Namespace, name}
 	t.Remove(revID)
 }
+
+type infiniteBreaker struct {
+	mu          sync.RWMutex
+	broadcast   chan struct{}
+	concurrency int32
+}
+
+func (ib *infiniteBreaker) Capacity() int {
+	return int(atomic.LoadInt32(&ib.concurrency))
+}
+
+func zeroOrOne(x int) int32 {
+	if x == 0 {
+		return 0
+	}
+	return 1
+}
+
+func (ib *infiniteBreaker) UpdateConcurrency(cc int) error {
+	rcc := zeroOrOne(cc)
+	// We lock here to make sure two scale up events don't
+	// stomp on each other's feet.
+	ib.mu.Lock()
+	defer ib.mu.Unlock()
+	old := atomic.SwapInt32(&ib.concurrency, rcc)
+
+	// Scale up/down event.
+	if old != rcc {
+		if rcc == 0 {
+			// Scaled to 0.
+			ib.broadcast = make(chan struct{})
+		} else {
+			close(ib.broadcast)
+		}
+	}
+	return nil
+}
+
+func (ib *infiniteBreaker) Maybe(ctx context.Context, thunk func()) bool {
+	has := ib.Capacity()
+	// We're scaled to serve.
+	if has > 0 {
+		thunk()
+		return true
+	}
+	select {
+	case <-ib.broadcast:
```

Comment:
This needs to be guarded by a mutex I think. You're writing to the variable concurrently. I had the race detector bark at me for usage like this. Doesn't harm the quick path though, which is great!

Reply:
Done.